File: apps/gpu-worker/src/timecode.py
"""Chuẩn hóa mốc thời gian về GIÂY (float) — dùng chung cho cả pipeline worker.

CC-1 (contract mismatch): client có thể gửi mốc thời gian ở dạng số (giây),
chuỗi số ("83.5"), hoặc timecode "HH:MM:SS" / "MM:SS" / "HH:MM:SS,mmm" (kiểu SRT).
Hai nơi trong worker tiêu thụ start/end:
  - translation_service: tính `duration = end - start` để căn lip-sync;
  - audio_engine.mix_audio: định vị clip bằng `int(start_s * 1000)`.
Cả hai trước đây dùng `float()` trần -> chuỗi "HH:MM:SS" ném ValueError rồi rơi về
fallback (pacing sai / bản mix hỏng). Gom về MỘT hàm để hành vi nhất quán, và để
client chỉ cần đối chiếu một hợp đồng duy nhất.

Không parse được -> 0.0 (KHÔNG đoán bừa). Client vẫn PHẢI gửi mốc thật; đây là
lớp phòng thủ, không phải chỗ để bịa dữ liệu.
"""


def to_seconds(value) -> float:
    """Đưa một mốc thời gian bất kỳ về giây (float). bool KHÔNG bị coi là 0/1 giây."""
    if isinstance(value, bool):
        return 0.0
    if isinstance(value, (int, float)):
        return float(value)
    if not value:
        return 0.0
    s = str(value).strip().replace(",", ".")
    try:
        return float(s)
    except ValueError:
        pass
    try:
        parts = [float(p) for p in s.split(":")]
    except ValueError:
        return 0.0
    sec = 0.0
    for p in parts:
        sec = sec * 60 + p
    return sec

File: apps/client/src/timecode.ts
/**
 * Quy mốc thời gian người dùng nhập ("HH:MM:SS", "MM:SS", "SS", hoặc số/thập phân
 * kiểu "1,5") về GIÂY (số thực) TRƯỚC KHI KÝ. Hợp đồng `ClientSegment.start/end`
 * là số giây; worker (translation_service tính duration, audio_engine căn mix) đọc
 * SỐ — nếu gửi thẳng chuỗi "HH:MM:SS", `float()` phía worker sẽ hỏng rồi rơi về
 * pacing mặc định sai (CC-1). Không parse được -> 0 (không đoán bừa). Logic này
 * phản chiếu 1-1 src/timecode.py phía worker để hai đầu nhất quán tuyệt đối.
 *
 * Tách khỏi App.tsx (Đợt 29) để unit-test được BẰNG hàm thuần, không kéo theo
 * React/Tauri vào module graph của test.
 */
export function timecodeToSeconds(tc: string): number {
  const s = (tc ?? '').trim().replace(',', '.');
  if (!s) return 0;
  const direct = Number(s);
  if (!Number.isNaN(direct)) return direct; // đã là số ("83", "1.5")
  const parts = s.split(':').map((p) => Number(p));
  if (parts.some((n) => Number.isNaN(n))) return 0; // rác -> 0
  return parts.reduce((acc, n) => acc * 60 + n, 0);
}
